src/D2L.CodeStyle.Analyzers/Immutability/ImmutableAttributeConsistencyChecker.cs

```diff
@@ -86,52 +86,88 @@ CancellationToken cancellationToken
-						baseTypeOrInterface: baseTypeInfo.Type,
-						compilation: m_compilation,
-						cancellationToken
-					);
-
-					m_diagnosticSink(
-						Diagnostic.Create(
-							Diagnostics.MissingTransitiveImmutableAttribute,
-							syntax.Identifier.GetLocation(),
-							properties: FixArgs,
-							typeInfo.Type.GetFullTypeName(),
-							baseTypeInfo.IsConditional ? " (or [ConditionallyImmutable])" : "",
-							isInterface ? "interface" : "base class",
-							baseTypeInfo.Type.GetFullTypeName()
-						)
-					);
-
-					break;
+				// If the base type is [Immutable] then the implementing type should be as well
+				case ((ImmutableTypeKind.Total, false), (ImmutableTypeKind.Total, true)):
+					RaiseMissingAttribute();
+					return;
 
-				default:
-					throw new NotImplementedException();
+				case ((ImmutableTypeKind.Total, true), (ImmutableTypeKind.Total, true)):
+					InspectConditionalParameterApplication();
+					return;
+			}
+
+			void RaiseMissingAttribute() {
+				(TypeDeclarationSyntax syntax, _) = typeInfo.Type.ExpensiveGetSyntaxImplementingType(
+					baseTypeOrInterface: baseTypeInfo.Type,
+					compilation: m_compilation,
+					cancellationToken
+				);
+
+				m_diagnosticSink(
+					Diagnostic.Create(
+						Diagnostics.MissingTransitiveImmutableAttribute,
+						syntax.Identifier.GetLocation(),
+						properties: FixArgs,
+						typeInfo.Type.GetFullTypeName(),
+						baseTypeInfo.IsConditional ? " (or [ConditionallyImmutable])" : "",
+						baseTypeInfo.Type.TypeKind == TypeKind.Interface ? "interface" : "base class",
+						baseTypeInfo.Type.GetFullTypeName()
+					)
+				);
+			}
+
+			void InspectConditionalParameterApplication() {
+				foreach( ITypeParameterSymbol typeParameter in typeInfo.ConditionalTypeParameters ) {
+					bool parameterUsed = false;
+					for( int i = 0; i < baseTypeInfo.Type.TypeArguments.Length && !parameterUsed; i++ ) {
+						ITypeSymbol typeArgument = baseTypeInfo.Type.TypeArguments[ i ];
+						if( !SymbolEqualityComparer.Default.Equals( typeParameter, typeArgument ) ) {
+							continue;
+						}
+
+						ITypeParameterSymbol baseTypeParameter = baseTypeInfo.Type.TypeParameters[ i ];
+						if( baseTypeInfo.ConditionalTypeParameters.Contains( baseTypeParameter, SymbolEqualityComparer.Default ) ) {
+							parameterUsed = true;
+							break;
+						}
+					}
+
+					if( !parameterUsed ) {
+						(TypeDeclarationSyntax syntax, _) = typeInfo.Type.ExpensiveGetSyntaxImplementingType(
+							baseTypeOrInterface: baseTypeInfo.Type,
+							compilation: m_compilation,
+							cancellationToken
+						);
+						m_diagnosticSink(
+							Diagnostic.Create(
+								Diagnostics.UnappliedConditionalImmutability,
+								syntax.Identifier.GetLocation(),
+								typeInfo.Type.GetFullTypeName(),
+								baseTypeInfo.Type.TypeKind == TypeKind.Interface ? "interface" : "base class",
+								baseTypeInfo.Type.GetFullTypeName()
+							)
+						);
+						return;

```

Comment:
We talked about how we could drop this return if the diagnostic was on `typeParameter` instead, like we could emit multiple diagnostics instead. I dno, consider it since this is a new diagnostic anyway